yesod-core/Yesod/Core/Handler.hs

```diff
@@ -787,6 +789,30 @@ setHeader :: MonadHandler m => Text -> Text -> m ()
 setHeader = addHeader
 {-# DEPRECATED setHeader "Please use addHeader instead" #-}
 
+-- | Replace an existing header with a new value or add a new header
+-- if not present.
+--
+-- Note that, while the data type used here is 'Text', you must provide only
+-- ASCII value to be HTTP compliant.
+--
+-- @since 1.4.36
+replaceOrAddHeader :: MonadHandler m => Text -> Text -> m ()
+replaceOrAddHeader a b =
+  let header = Header (encodeUtf8 a) (encodeUtf8 b)
+  in modify $ \g -> g {ghsHeaders = replaceHeader header (ghsHeaders g)}
+  where
+    sameHeaderName :: Header -> Header -> Bool
+    sameHeaderName (Header n1 _) (Header n2 _) = n1 == n2
+    sameHeaderName _ _ = False
+
+    replaceHeader :: Header -> Endo [Header] -> Endo [Header]
+    replaceHeader header endo =
+      let allHeaders :: [Header] = appEndo endo []
+      in Endo
```

Comment:
@snoyberg 

> The structure here looks wrong, you're adding the new header to the front of the list, then any additional headers added after, and finally the previously defined header. 

I was under the impression that the ordering didn't matter. I have updated the code to fix it and have added additional test to ensure that any header isn't lost because of the above function.